zeroknowledgeproof/src/main/java/com/zeroknowledgeproof/rangeProof/RangeProofProver.kt

```diff
@@ -0,0 +1,199 @@
+
+        // Publicly give a proof that two integers are equal
+        val cDPrime = (cPrime.modPow(w * w, N) * h.modPow(rDPrime, N)).mod(N) // The commitment of w in base cPrime
+        val cDPrimeIsSquare = proveCommittedNumberIsSquare(w, rDPrime, cPrime, h, cDPrime)
+        println(">STEP 3 FINISHED")
+
+        // Step 4: Generate m1, m2, m3
+        val sum = w * w * toBigInt((m - a + 1) * (b - m + 1))
+        val (m1, m2, m4) = calculateMValues(sum)
+        val m3 = m4 * m4
+
+        // Generate r1, r2, r3 such that their sum is equal to w^2 (( b - m + 1)r + r') + r''
+        val (r1, r2, r3) = calculateRValues(w * w * (toBigInt(b - m + 1) * r + rPrime) + rDPrime)
+        // Then, prove that m3 is a square.
+        val c1Prime = (g.modPow(m1, N) * h.modPow(r1, N)).mod(N)
+        val c2Prime = (g.modPow(m2, N) * h.modPow(r2, N)).mod(N)
+        val c3Prime = (cDPrime * calculateInverse(c1Prime * c2Prime, N)).mod(N)
+        val m3IsSquare = proveCommittedNumberIsSquare(m4, r3, g, h, c3Prime)
+
+        println(">STEP 4 FINISHED")
+
+        // Put all results in the result struct, return it.
+        val pubRes = SetupPublicResult(g = g, h = h, c = c, c1 = c1, c2 = c2, cPrime = cPrime,
+                cDPrime = cDPrime, c1Prime = c1Prime, c2Prime = c2Prime, c3Prime = c3Prime,
+                sameCommitment = proofEqualIntegers, cDPrimeIsSquare = cDPrimeIsSquare,
+                m3IsSquare = m3IsSquare, k1 = BigInteger(2048, rand))
+        // To perform the verification, the verifier needs a bound. This bound is k1, generated by the prover.
+
+        // Save the private parameters, that the verifier should NOT obtain
+        val privRes = SetupPrivateResult(m1 = m1, m2 = m2, m3 = m3, r1 = r1, r2 = r2, r3 = r3)
+        return Pair(pubRes, privRes)
+    }
+
+    fun answerUniqueChallenge(privSetup: SetupPrivateResult, challenge: Challenge): InteractivePublicResult {
+        // Take values from relative structs
+        // private values
+        val (m1, m2, m3, r1, r2, r3) = privSetup
+
+        // Step 5: Generate s,t in Zk1 - {0} //
+        // Already generated and passed to us
+
+        val s = challenge.s
+        val t = challenge.t
+
+        // Step 6: NumberProofProver publishes x, y, u, v
+        val x = s * m1 + m2 + m3
+        val y = m1 + t * m2 + m3
+
+        val u = s * r1 + r2 + r3
+        val v = r1 + t * r2 + r3
+
+        return InteractivePublicResult(x = x, y = y, u = u, v = v, challenge = challenge)
+    }
+
+    fun proveTwoCommittedIntegersAreEqual(committedNum: BigInteger, r1: BigInteger, r2: BigInteger,
+                                          g1: BigInteger, h1: BigInteger, g2: BigInteger, h2: BigInteger,
+                                          y1: BigInteger, y2: BigInteger): CommittedIntegerProof {
+
+        // Init security parameters
+        val b = BigInteger(512, SecureRandom())
+
+
+        //TODO: Hash function H should output 2t-bit strings.
```

Comment:
is still still a todo or is it solved by the sha-512 down below?

Reply:
We can use SHA256. That'll reduce it to 256 bits, which is still more than "necessary" but I'd rather not use SHA1.

Reply:
ah so the problem is that it's outputting too many bits right now?